apps/src/templates/ShowCodeToggle.js

```diff
@@ -0,0 +1,97 @@
+import React, {Component, PropTypes} from 'react';
+import logToCloud from '../logToCloud';
+import {singleton as studioApp} from '../StudioApp';
+import {PaneButton} from './PaneHeader';
+import msg from '@cdo/locale';
+
+const BLOCKS_GLYPH_LIGHT = "data:image/gif;base64,R0lGODlhEAAQAIAAAP///////yH+GkNyZWF0ZWQgd2l0aCBHSU1QIG9uIGEgTWFjACH5BAEKAAEALAAAAAAQABAAAAIdjI+py40AowRp2molznBzB3LTIWpGGZEoda7gCxYAOw==";
+const BLOCKS_GLYPH_DARK = "data:image/gif;base64,R0lGODlhEAAQAIAAAE1XX01XXyH+GkNyZWF0ZWQgd2l0aCBHSU1QIG9uIGEgTWFjACH5BAEKAAEALAAAAAAQABAAAAIdjI+py40AowRp2molznBzB3LTIWpGGZEoda7gCxYAOw==";
+
+const styles = {
+  blocksGlyph: {
+    display: 'none',
+    height: 18,
+    lineHeight: '24px',
+    verticalAlign: 'text-bottom',
+    paddingRight: 8
+  },
+  blocksGlyphRtl: {
+    paddingRight: 0,
+    paddingLeft: 8,
+    transform: 'scale(-1, 1)',
+    MozTransform: 'scale(-1, 1)',
+    WebkitTransform: 'scale(-1, 1)',
+    OTransform: 'scale(-1, 1)',
+    msTransform: 'scale(-1, 1)',
+  },
+};
+
+export default class ShowCodeToggle extends Component {
+  static propTypes = {
+    hasFocus: PropTypes.bool,
+    isRtl: PropTypes.bool,
+    isMinecraft: PropTypes.bool,
+    onToggle: PropTypes.func,
+  };
+
+  state = {
+    showingCode: false,
+  };
+
+  onClick = () => {
+    if (studioApp.editCode) {
+      // are we trying to toggle from blocks to text (or the opposite)
+      let fromBlocks = studioApp.editor.currentlyUsingBlocks;
+
+      let result;
+      try {
+        result = studioApp.editor.toggleBlocks();
+      } catch (err) {
+        result = {error: err, nonDropletError: true};
+      }
+      if (result && result.error) {
+        logToCloud.addPageAction(logToCloud.PageAction.DropletTransitionError, {
+          dropletError: !result.nonDropletError,
+          fromBlocks,
+        });
+        studioApp.showToggleBlocksError();
+      }
+      studioApp.onDropletToggle();
+      this.setState({showingCode: !studioApp.editor.currentlyUsingBlocks});
+    } else {
+      studioApp.showGeneratedCode();
```

Comment:
It looks like the two cases that aren't covered by tests here are:
1. `studioApp.editor.toggleBlocks()` throws an error (not sure when/why that would happen, but it's clear we planned for it).
2. Non-editCode levels (that is, using Blockly instead of Droplet).

I'm a little surprised that our integration tests don't cover the latter.  Have you manually tested that case?